[SNAP-2366] row buffer fault-in, forced rollover, merge small batches

sumwale opened the pull request:
## Changes proposed in this pull request

- add check for the two cases in table stats service:
  - "large enough" row buffer (currently "large enough" is anything more than maxDeltaRows/8)
    that has not seen any updates/deletes since the last check; in this case schedule
    a task to force rollover the row buffer in column table
  - also check if a bucket of column table has multiple small batches (non-transactional check);
    if so then submit a task to merge those after checking for transactional snapshot;
    merge is done by locally created ColumnTableScan->ColumnInsertExec plan where the scan
    uses an iterator only on the small batches
- added a ColumnFormatStatsIterator that can take a bunch of stats rows and create an iterator
  over just those (like required for batch merge)
- added new scan metrics for disk reads: a) disk rows from row buffer,
  b) partial column batches on disk, c) full column batches on disk
- extended SQLMetrics types with a new SPLIT_SUM_METRIC that allows displaying multiple
  metrics against a common name; ColumnTableScan now uses this to combine some metrics
  else it becomes too large in display (especially for the newly added disk read metrics)
- use hive-metadata (ExternalTableMetaData) to get number of rows instead of getting
  from row buffer table (that is subject to changes in future)
- added a metric for remote batch fetch
- fix an NPE in SnappyTableStatsProviderService while filling up
  result map from members since CHM cannot hold null values
- use a common entry map in ColumnFormatIterator disk iteration instead of creating
  separate for every column batch
- added implementation of PURGE_CODEGEN_CACHES in StoreCallbacksImpl
- limit to one task per table for background rolloverRowBuffer and mergeSmallBatches tasks

## Patch testing

precheckin; new unit tests to be added next

## ReleaseNotes.txt changes

NA

## Other PRs 

https://github.com/SnappyDataInc/snappy-store/pull/391

Commit:
Update store link

Commit:
fix for SNAP-2365

Commit:
[SNAP-2366] row buffer fault-in, forced rollover, merge small batches

- add check for the two cases in table stats service:
  - "large enough" row buffer (currently "large enough" is anything more than maxDeltaRows/8)
    that has not seen any updates/deletes since the last check; in this case schedule
    a task to force rollover the row buffer in column table
  - also check if a bucket of column table has multiple small batches (non-transactional check);
    if so then submit a task to merge those after checking for transactional snapshot;
    merge is done by locally created ColumnTableScan->ColumnInsertExec plan where the scan
    uses an iterator only on the small batches
- added a ColumnFormatStatsIterator that can take a bunch of stats rows and create an iterator
  over just those (like required for batch merge)
- added new scan metrics for disk reads: a) disk rows from row buffer,
  b) partial column batches on disk, c) full column batches on disk
- extended SQLMetrics types with a new SPLIT_SUM_METRIC that allows displaying multiple metrics
  against a common name; ColumnTableScan now uses this to combine some metrics else it becomes
  too large in display (especially for the newly added disk read metrics)
- use hive-metadata (ExternalTableMetaData) to get number of rows instead of getting
  from row buffer table (that is subject to changes in future)

Commit:
update store link

Commit:
Some optimizations and fixed few issues

- fixed disk metrics collection added previously; set the metric correctly
  for both row buffer iterator (ResultSetTraversal) and ColumnFormatIterator
- added a metric for remote batch fetch
- fixed multiple ColumnTableScans causing split metrics to add up into one
  ColumnTableScan; now use a unique ID for split metrics for each ColumnTableScan instance
- fix an NPE in SnappyTableStatsProviderService while filling up
  result map from members since CHM cannot hold null values
- use a common entry map in ColumnFormatIterator disk iteration instead of creating
  separate for every column batch
- added implementation of PURGE_CODEGEN_CACHES as StoreCallbacksImpl.clearCodegenCaches
- limit to one task per table for background rolloverRowBuffer and mergeSmallBatches tasks
- replaced a few usage of Map.put with justPut for koloboke maps

Commit:
minor change

Commit:
Fixing precheckin failure in SNAP-2365 and and adding similar fix for replicated table

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Conflicts:
	core/src/main/scala/io/snappydata/SnappyTableStatsProviderService.scala
	core/src/main/scala/io/snappydata/TableStatsProviderService.scala
	store

Commit:
Merge remote-tracking branch 'origin/SNAP-2365' into SNAP-2366

Conflicts:
	cluster/src/test/scala/io/snappydata/benchmark/TPCHColumnPartitionedTable.scala

rishitesh commented on core/src/main/scala/io/snappydata/SnappyTableStatsProviderService.scala:
```diff
@@ -215,55 +221,265 @@ object SnappyEmbeddedTableStatsProviderService extends TableStatsProviderService
+              partitionColumnAliases = Nil, baseRelation = null, schema, allFilters = Nil,
+              schemaAttrs, caseSensitive = true)
+            // reduce min delta row size to avoid going through rolloverRowBuffers again
+            val insertPlan = ColumnInsertExec(tableScan, Nil, Nil,
```
Can you please explain here, maybe in the comment section how this will handle updated columns and deleted columns?

sumwale replied:
The ColumnBatchIterator is passed the stats row entry. Rest all the columns, including delta/delete are looked up by the iterator (see ColumnFormatStatsIterator.getColumnValue) when the generated code asks for them.

So this will be same as iterating batches in ColumnTableScan that will return merged entries with deltas/deletes applied. The ColumnInsert is tied to output of this hence will create a combined merged batch.

Note that merging of deltas into main batch (when they become large enough) or deletes into main batch (when large number of entries are deleted) will be handled separately. This does not depend on main batch being small rather deltas being large. That merge needs to be done in the operation thread itself (that has created the last delta causing it to grow large). Right now only the case where all entries are deleted is handled.

rishitesh reviewed (commented):
First set of comments. Can see more after one of the points about updated & deleted columns is clarified.

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Conflicts:
	core/src/main/scala/org/apache/spark/sql/store/CompressionCodecId.scala

sumwale commented on core/src/main/scala/io/snappydata/SnappyTableStatsProviderService.scala:
```diff
@@ -215,55 +221,265 @@ object SnappyEmbeddedTableStatsProviderService extends TableStatsProviderService
+    }
+  }
+
+  private def withExceptionHandling(f: => Unit, doFinally: () => Unit = null): Unit = {
```
Will do

sumwale commented on core/src/main/scala/io/snappydata/SnappyTableStatsProviderService.scala:
```diff
@@ -215,55 +221,265 @@ object SnappyEmbeddedTableStatsProviderService extends TableStatsProviderService
+        logInfo(s"Found small batches in ${pr.getName}: ${smallBatches.map(_.getId)}")
+        val cache = pr.getGemFireCache
+        implicit val executionContext = Utils.executionContext(cache)
+        Future(withExceptionHandling({
```
That cannot happen because this deliberately uses a "mergeTasks" map with "computeIfAbsent" that will add only one future for a PR at a time. The mergeTasks is cleared for that PR only at the end of Future execution. Same for rolloverTasks.

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Commit:
coarse container locking to fix rollover/merge when running in parallel with update/delete

Commit:
temp

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Commit:
minor formatting change

Commit:
fixes

Commit:
fix a ClassCast

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Commit:
fixing putInto which clears context prematurely

added an OperationContext in SnappySession that can be used to persist context across
multiple plan executions (e.g. caching for putInto, then actual execution)

Commit:
fix COMMIT call

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Commit:
update store link

Commit:
updates and fixes

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Commit:
fix build issues after master merge

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Commit:
minor cleanups

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Commit:
minor updates to tests

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

Commit:
Merge remote-tracking branch 'origin/master' into SNAP-2366

PradeepSurale commented:
Can one of the admins verify this patch?
